### PR TITLE
Issue license to holder

### PR DESCRIPTION
**Pull request by jonmattgray**
- [x] Create a license credential definition if one doesn't exist
- [x] Listens for license presentation proof proposals from a `holder`
- [x] Send a presentation proof to `holder` upon receiving a proposal. Deletes the proposal
- [x] Issues licenses for verified proofs (proof presentation and verification happens automatically with `ACAPY_AUTO_` envs on cloudagents). Deletes proof record

### CODE REVIEW

File: src/interface/hooks/use-get-looped-present-proof-records.js
Comment (mattdean-digicatapult) [inline src/interface/hooks/use-get-looped-present-proof-records.js]: That looks like a magic number to me. Why 3s?

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (mattdean-digicatapult) [inline src/components/AgentAuthority/ContentWrap/ContentWrap.js]: This is pretty yucky. So each of these components (and their subcomponents) is handling a different part of the flow on loops in a disconnected way. And then they each render their own Errors completely separate from one another right? I wonder if we could handle this better in a single hook that then did the whole flow and reported an amalgameted error/status back up?

File: src/components/AgentAuthority/IssueLicense/IssueLicense.js
Comment (mattdean-digicatapult) [inline src/components/AgentAuthority/IssueLicense/IssueLicense.js]: nice, so much cleaner

File: src/components/AgentAuthority/IssueLicense/IssueLicense.js
Comment (mattdean-digicatapult) [inline src/components/AgentAuthority/IssueLicense/IssueLicense.js]: This is um ... interesting. I get it's not your fault @jonmattgray

**Review by mattdean-digicatapult — COMMENTED** [review]
I think we can make some further improvements on the flow here. There's a bunch of logic happening across components which I'm not wild about. Let me know if you want to talk it through

File: src/interface/hooks/use-get-looped-present-proof-records.js
Comment (jonmattgray) [reply src/interface/hooks/use-get-looped-present-proof-records.js]: I'll set it back to 1s

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (jonmattgray) [reply src/components/AgentAuthority/ContentWrap/ContentWrap.js]: Yep `RequestProof` polls for proposals and `IssueLicense` polls for verified proofs

Comment (jonmattgray) [reply src/components/AgentAuthority/ContentWrap/ContentWrap.js]: changed to mostly use a single hook (apart from polling for proposals) with callbacks

File: src/components/AgentAuthority/IssueLicense/IssueLicense.js
Comment (jonmattgray) [reply src/components/AgentAuthority/IssueLicense/IssueLicense.js]: empty functions like this are gone now

File: package.json
Comment (n3op2) [inline package.json]: maybe we could use promise and save on the module?

```js
await new Promise(r => setTImeout(r, 1000))
```

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (n3op2) [inline src/components/AgentAuthority/ContentWrap/ContentWrap.js:14]: since we are using the left/right columns approach maybe this should be on either or rename columns to panel and middle have as a context? just a thought

File: src/interface/hooks/use-issue-license.js
Comment (n3op2) [inline src/interface/hooks/use-issue-license.js:96]:
```js
await new Promise(r => setTimeout(r, 3000))
```

File: src/interface/hooks/use-issue-license.js
Comment (n3op2) [inline src/interface/hooks/use-issue-license.js:149]: It would be nice to refactor hooks to use a shared object for fetching state along with error/success

File: src/interface/hooks/use-post-credential-definitions.js
Comment (n3op2) [inline src/interface/hooks/use-post-credential-definitions.js]:
```js
str.replace(/\s+/g, '_')
```

File: src/interface/hooks/use-post-credential-definitions.js
Comment (n3op2) [inline src/interface/hooks/use-post-credential-definitions.js]:
```js
schemaDefTagName = schemaDefName.replace(/\s+/g, '_')
```

would reduce the number of lines and I don't think it requires a helper function

File: src/components/AgentAuthority/LicenseCredDef/LicenseCredDef.js
Comment (mattdean-digicatapult) [inline src/components/AgentAuthority/LicenseCredDef/LicenseCredDef.js]: I still don't like the duplicate `Error` component here and on [src/components/AgentAuthority/IssueLicense/IssueLicense.js](https://github.com/digicatapult/veritable-authority/pull/4/files#diff-f315495d2dae22e054082d8820f1d2dd9e729e345b01234e437154775c59efdf)

File: src/interface/hooks/use-post-issue-credential-send-offer.js
Comment (n3op2) [inline src/interface/hooks/use-post-issue-credential-send-offer.js]: sounds like a good candidate for reduce

**Review by mattdean-digicatapult — CHANGES_REQUESTED** [review]
This is looking much better. There's still a bit more cleanup that I'll pickup today @jonmattgray.

File: src/interface/hooks/use-post-present-proof-send-request.js
Comment (n3op2) [inline src/interface/hooks/use-post-present-proof-send-request.js]: what does it mean? `Prs4zk`?

File: src/interface/hooks/use-post-present-proof-send-request.js
Comment (n3op2) [inline src/interface/hooks/use-post-present-proof-send-request.js]: this looks a like a good place for bug to happen when this is a srtring and we perform some string methods and later treating like a number. If JS magic for some reason would fail or would not work, this could turn into a bug.

File: src/interface/hooks/use-post-schemas.js
Comment (n3op2) [inline src/interface/hooks/use-post-schemas.js]: I think we should remove this comment. there was a request from other repos

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (mattdean-digicatapult) [reply src/components/AgentAuthority/ContentWrap/ContentWrap.js:14]: sorry @n3op2 not sure I understand this one?

File: package.json
Comment (mattdean-digicatapult) [reply package.json]: We've used `delay` in other places/projects but sure we can remove

File: src/interface/hooks/use-issue-license.js
Comment (mattdean-digicatapult) [reply src/interface/hooks/use-issue-license.js:149]: Maybe. Normally I'd expect a context component higher up the stack (like Auth0 does). Then that would be able to report API success error generically. I think any sort of major refactor like that has to be out of scope for this work

File: src/interface/hooks/use-post-present-proof-send-request.js
Comment (mattdean-digicatapult) [reply src/interface/hooks/use-post-present-proof-send-request.js]: not sure, will investigate

File: src/interface/hooks/use-post-present-proof-send-request.js
Comment (mattdean-digicatapult) [reply src/interface/hooks/use-post-present-proof-send-request.js]: Agreed this is ugly. Will find a better way

File: src/components/AgentAuthority/LicenseCredDef/LicenseCredDef.js
Comment (mattdean-digicatapult) [reply src/components/AgentAuthority/LicenseCredDef/LicenseCredDef.js]: Updated

File: src/interface/hooks/use-post-present-proof-send-request.js
Comment (mattdean-digicatapult) [reply src/interface/hooks/use-post-present-proof-send-request.js]: it's just a `pr`edicate`s` `for` `z`ero `k`nowledge. Renamed to `predicates`

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (n3op2) [reply src/components/AgentAuthority/ContentWrap/ContentWrap.js:14]: we wrap all the content in column right or left, however we have added this before other wrappers, so thought maybe it would make sense to keep in one of the columns of create another wrapper either instead or besides

File: src/interface/hooks/use-issue-license.js
Comment (n3op2) [reply src/interface/hooks/use-issue-license.js:149]: yea, i like your idea

**Review by mattdean-digicatapult — APPROVED** [review]
Approved, but since I'm a co-author waiting for approval from @n3op2 before merging

File: src/components/AgentAuthority/ContentWrap/ContentWrap.js
Comment (mattdean-digicatapult) [reply src/components/AgentAuthority/ContentWrap/ContentWrap.js:14]: got it. I think the issue is it won't fit neatly into either and the reality is this should be on a backend. I'm happy leaving it like this as all it does is render an error (if there is one) and we'll remove it when this moves to the backend

**Review by n3op2 — APPROVED** [review]

**Pull request merged**
